plutus-metatheory/src/VerifiedCompilation/UInline.lagda.md

```diff
@@ -0,0 +1,111 @@
+open import VerifiedCompilation.UntypedViews using (Pred; isCase?; isApp?; isLambda?; isForce?; isBuiltin?; isConstr?; isDelay?; isTerm?; allTerms?; iscase; isapp; islambda; isforce; isbuiltin; isconstr; isterm; allterms; isdelay)
+open import VerifiedCompilation.UntypedTranslation using (Translation; translation?; Relation; convert; reflexive)
+import Relation.Binary as Binary using (Decidable)
+open import Untyped.RenamingSubstitution using (_[_])
+open import Agda.Builtin.Maybe using (Maybe; just; nothing)
+open import Untyped using (_⊢; case; builtin; _·_; force; `; ƛ; delay; con; constr; error)
+open import Relation.Nullary using (Dec; yes; no; ¬_)
+open import Untyped.RenamingSubstitution using (weaken)
+open import Data.Empty using (⊥)
+import Relation.Binary.PropositionalEquality as Eq
+open Eq using (_≡_; refl)
+
+```
+## Translation Relation
+
+Abstractly, inlining is much like β-reduction - where a term is applied to a lambda,
+the term is substituted in. This can be expressed quite easily and cleanly with the
+substitution operation from the general metatheory.
+
+```
+data pureInline : Relation where
+  tr : {X : Set} {{_ : DecEq X}} {x x' : X ⊢} → Translation pureInline x x' → pureInline x x'
+  _⨾_ : {X : Set} {{_ : DecEq X}} {x x' x'' : X ⊢} → pureInline x x' → pureInline x' x'' → pureInline x x''
+  inline : {X : Set} {{_ : DecEq X}}{x' : X ⊢} {x : Maybe X ⊢} {y : X ⊢}
+         → pureInline (x [ y ]) x'
+         → pureInline (ƛ x · y) x'
+
+_ : pureInline {⊥} (ƛ (` nothing) · error) error
+_ = inline (tr reflexive)
+
+_ : {X : Set} {a b : X} {{_ : DecEq X}} → pureInline (((ƛ (ƛ ((` (just nothing)) · (` nothing)))) · (` a)) · (` b)) ((` a) · (` b))
+_ = tr (Translation.app (Translation.istranslation (inline (tr reflexive))) reflexive) ⨾ inline (tr reflexive)
+
+```
+However, this has several intermediate values that are very hard to determine for a decision procedure.
+
+The Haskell code works by building an environment of variable values and then inlines from these. We can
+replicate that here to try and make the decision procedure easier.
+```
+data Env {X : Set}{{_ : DecEq X}} : Set where
+  □ : Env
+  _,_ : Env {X} → (X ⊢) → Env {X}
+
+```
+# Decidable Inline Type
+
+This type is closer to how the Haskell code works, and also closer to the way Jacco's inline example works in the literature.
+
+It recurses to the Translation type, but it needs to not do that initially or the `translation?` decision procedure
+will recurse infinitely. Instead there is the `last-sub` constructor to allow the recursion only if at least
+something has happened.
+
+```
+
+data Inline {X : Set} {{ _ : DecEq X}} : Env {X} → (X ⊢) → (X ⊢) → Set₁ where
+  var : {x y z : X ⊢} {e : Env} → Inline (e , x) z y → Inline e (z · x) y
+  last-sub : {x : (Maybe X) ⊢ } {y v : X ⊢} → Translation (Inline □) (x [ v ]) y → Inline (□ , v) (ƛ x) y
+  sub : {x : (Maybe X) ⊢ } {y v v₁ : X ⊢} {e : Env} → Inline (e , v₁) (x [ v ]) y → Inline ((e , v₁) , v) (ƛ x) y
```

Comment:
Once we start testing performance of the certifier, this pass should be our first, because what I'm seeing here is extremely inefficient and I'd be surprised to learn that it can handle real-world UPLC scripts. Although maybe Agda being non-strict saves some of the trouble.

Reply:
I think you can avoid the multiple parenthesis in `((e , v₁) , v)` if you add a `infixl` declaration: https://agda.readthedocs.io/en/v2.7.0.1/language/mixfix-operators.html